README.md
# GroupListView package for Flutter.

[![pub package](https://img.shields.io/pub/v/group_list_view.svg)](https://pub.dev/packages/group_list_view)

A ListView that allows you to group list items and support headers like iOS UITableView section.

<img src="https://raw.githubusercontent.com/Daniel-Ioannou/flutter_group_list_view/master/assets/ReadMe%20%20Screenshot.png" width="300"> 

### Features
* List Items can be grouped.
* Support headers for each group.
* All fields from `ListView.builder` constructor available.

## Getting Started

 Add the package to your pubspec.yaml:

 ```yaml
 group_list_view: ^1.1.1
 ```
 
 In your dart file, import the library:

 ```Dart
import 'package:group_list_view/group_list_view.dart';
 ``` 
 
 Instead of using a `ListView` create a `GroupListView` Widget:
 
 ```Dart
  Map<String, List> _elements = {
    'Team A': ['Klay Lewis', 'Ehsan Woodard', 'River Bains'],
    'Team B': ['Toyah Downs', 'Tyla Kane'],
  };
  
  GroupListView(
    sectionsCount: _elements.keys.toList().length,
    countOfItemInSection: (int section) {
      return _elements.values.toList()[section].length;
    },
    itemBuilder: (BuildContext context, IndexPath index) {
      return Text(
        _elements.values.toList()[index.section][index.index],
        style: TextStyle(color: Colors.white, fontSize: 18),
      );
    },
    groupHeaderBuilder: (BuildContext context, int section) {
      return Padding(
        padding: const EdgeInsets.symmetric(horizontal: 15, vertical: 8),
        child: Text(
          _elements.keys.toList()[section],
          style: TextStyle(fontSize: 18, fontWeight: FontWeight.w600),
        ),
      );
    },
    separatorBuilder: (context, index) => SizedBox(height: 10),
    sectionSeparatorBuilder: (context, section) => SizedBox(height: 10),
  );
```

### Parameters:
* `sectionsCount`: The number of sections in the ListView. (required)
* `countOfItemInSection`. Function which returns the number of items(rows) in a specified section. (required)
* `itemBuilder`: Function which returns an Widget which defines the item at the specified `IndexPath`. `itemBuilder` provides the current section and index. (required)
```Dart
  Widget _itemBuilder(BuildContext context, IndexPath index) {
     return Text('${_elements[index.section][index.index]}');
  }
```  
* `groupHeaderBuilder`: Function which returns an Widget which defines the section header for each group. (required)
* `separatorBuilder`: Function which returns an Widget which defines the divider/separator at the specified IndexPath.

  The `itemBuilder` and `separatorBuilder` callbacks should actually create widget instances when called. 

## Contributions
Contributions of any kind are more than welcome! Feel free to fork and improve country_code_picker in any way you want, make a pull request, or open an issue.


CHANGELOG.md
## [1.1.1] - 10 Mar 2021

* Stable null-safe release

## [1.1.0] - 02 Mar 2021

* Migrated to null safety

## [1.0.6] - 21 Jul 2020

* Support `sectionSeparatorBuilder` 
  
  Separators only appear between sections: separator 0 appears after section 0 and the last separator appears after the last section.

```Dart
Widget sectionSeparatorBuilder(BuildContext context, int section) {
   return SizedBox(height: 10);
}
```  

## [1.0.5] - 04 Jun 2020

* Add assert for all the required parameters.
* Update documentation.
* Setting up Lint Rules.

## [1.0.4] - 20 May 2020

* Support `separatorBuilder`  
  
  Separators only appear between list items: separator 0 appears after item 0 and the last separator appears after the last item.

```Dart
Widget separatorBuilder(BuildContext context, IndexPath index) {
   return SizedBox(height: 10);
}
```  

## [1.0.3] - 12 May 2020

* Improve build performance.

## [1.0.2] - 10 May 2020

* Update documentation.

## [1.0.1] - 08 May 2020

* Support all fields from `ListView.builder` constructor.

## [1.0.0] - 06 May 2020

* Initial developers preview release.


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:group_list_view/group_list_view.dart';
import 'package:group_listview_example/app_colors.dart';

void main() => runApp(MyApp());

Map<String, List> _elements = {
  'Team A': ['Klay Lewis', 'Ehsan Woodard', 'River Bains'],
  'Team B': ['Toyah Downs', 'Tyla Kane'],
  'Team C': ['Marcus Romero', 'Farrah Parkes', 'Fay Lawson', 'Asif Mckay'],
  'Team D': [
    'Casey Zuniga',
    'Ayisha Burn',
    'Josie Hayden',
    'Kenan Walls',
    'Mario Powers'
  ],
  'Team Q': ['Toyah Downs', 'Tyla Kane', 'Toyah Downs'],
};

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Group List View Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        appBar: AppBar(
          title: Text('Group List View Demo'),
        ),
        body: GroupListView(
          sectionsCount: _elements.keys.toList().length,
          countOfItemInSection: (int section) {
            return _elements.values.toList()[section].length;
          },
          itemBuilder: _itemBuilder,
          groupHeaderBuilder: (BuildContext context, int section) {
            return Padding(
              padding: const EdgeInsets.symmetric(horizontal: 15, vertical: 8),
              child: Text(
                _elements.keys.toList()[section],
                style: TextStyle(fontSize: 18, fontWeight: FontWeight.w600),
              ),
            );
          },
          separatorBuilder: (context, index) => SizedBox(height: 10),
          sectionSeparatorBuilder: (context, section) => SizedBox(height: 10),
        ),
      ),
    );
  }

  Widget _itemBuilder(BuildContext context, IndexPath index) {
    String user = _elements.values.toList()[index.section][index.index];
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: 8.0),
      child: Card(
        elevation: 8,
        child: ListTile(
          contentPadding:
              const EdgeInsets.symmetric(horizontal: 18, vertical: 10.0),
          leading: CircleAvatar(
            child: Text(
              _getInitials(user),
              style: TextStyle(color: Colors.white, fontSize: 18),
            ),
            backgroundColor: _getAvatarColor(user),
          ),
          title: Text(
            _elements.values.toList()[index.section][index.index],
            style: TextStyle(fontSize: 16, fontWeight: FontWeight.w400),
          ),
          trailing: Icon(Icons.arrow_forward_ios),
        ),
      ),
    );
  }

  String _getInitials(String user) {
    var buffer = StringBuffer();
    var split = user.split(" ");
    for (var s in split) buffer.write(s[0]);

    return buffer.toString().substring(0, split.length);
  }

  Color _getAvatarColor(String user) {
    return AppColors
        .avatarColors[user.hashCode % AppColors.avatarColors.length];
  }
}
